File: 백준/Silver/1012. 유기농 배추/유기농 배추.cc
#include <iostream>
#include <queue>
#define MAX 51
using namespace std;
int graph[MAX][MAX];
int x_dir[4] = {1, -1, 0, 0};
int y_dir[4] = {0, 0, 1, -1};
int M, N;

void BFS(int i, int j){
    queue<pair<int, int>> q;
    pair<int, int> loc;
    loc.first = i;
    loc.second = j;
    q.push(loc);
    while(!q.empty()){
        int x = q.front().first;
        int y = q.front().second;
        q.pop();
        for(int d = 0; d < 4; d++){
            int next_x = x + x_dir[d];
            int next_y = y + y_dir[d];
            if ( 0 <= next_x < M and 0 <= next_y < N ){
                if ( graph[next_x][next_y] == 1 ){
                    graph[next_x][next_y] = 0;
                    loc.first = next_x;
                    loc.second = next_y;
                    q.push(loc);
                }
            }
        }
    }
}

int main(){
    int T, K, x, y, cnt;
    cin >> T;
    for (int t = 0; t < T; t++){
        cnt = 0;
        cin >> M >> N >> K;

        //그래프 초기화
        for (int m = 0; m < M; m++){
            for (int n = 0; n < N; n++){
                graph[m][n] = 0;
            }
        }

        //입력
        for (int k = 0; k < K; k++){
            cin >> x >> y;
            graph[x][y] = 1;
        }

        //그래프
        for (int m = 0; m < M; m++){
            for (int n = 0; n < N; n++){
                if ( graph[m][n] == 1){
                    BFS(m, n);
                    cnt++;
                }
            }
        }
        cout << cnt << endl;
    }

    return 0;
}

File: 백준/Silver/1012. 유기농 배추/유기농 배추.py
import sys
sys.setrecursionlimit(10**6)
from collections import deque

input = sys.stdin.readline
x_dir = [1, -1, 0, 0]
y_dir = [0, 0, 1, -1]
def DFS(i, j):
    cnt = 1
    graph[i][j] = 0
    for d in range(4):
        x = i + x_dir[d]
        y = j + y_dir[d]
        if 0 <= x and x < M and 0 <= y and y < N:
            if graph[x][y] == 1:
                cnt += DFS(x, y)
    return cnt

#def BFS():

T = int(input())
for _ in range(T):
    M, N, K = map(int, input().split())
    graph = [[0 for _ in range(N)]for _ in range(M)]
    for _ in range(K):
        i, j = map(int, input().split())
        graph[i][j] = 1
    num = 0
    for m in range(M):
        for n in range(N):
            if graph[m][n] == 1:
                DFS(m, n)
                num += 1
    print(num)